### PR TITLE
locations.geojson POC phase2: End-to-end partial support of json data

### PR DESCRIPTION
**Pull request by jcpitre**
**Summary:**

Builds on the poc written by @davidgamez (#1805), related to locations.geojson

What it does:
- Add the necessary classes to process locations.geojson (Classes with `Geojson` in the name)
- Read parts of the file
- Add a sample validator (GtfsGeojsonFeatureUniqueLocationIdValidator) and notice (UniqueLocationIdViolationNotice)

Not covered:

- Reading and processing geometry data.
- Processing the other files from Flex
- Adding other validators or notices apart from the sample one.

To be done:
- Add tests
- Comments

This is **NOT** ready to merge yet.

I modified an existing dataset that has flex data to create an error where the id of a feature in `locations.geojson` is the same as an id in `stops.txt`. 
[browncounty-mn-us--flex-v2-broken.zip](https://github.com/user-attachments/files/16871358/browncounty-mn-us--flex-v2-broken.zip)

Here is a capture of the sample notice in the report:
image

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Add or update any needed [documentation](https://github.com/MobilityData/gtfs-validator/tree/master/docs) to the repo 
- [ ] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsContainer.java
Comment (qcdyx) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsContainer.java:1]: Could you add class comments to help understanding the purpose of this abstract class?

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsDescriptor.java
Comment (qcdyx) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsDescriptor.java:1]: Same here, class comment could be useful.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/TableStatus.java
Comment (qcdyx) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/table/TableStatus.java:1]: What's the benefit of pulling TableStatus out from GtfsTableContainer? Just curious, coz nothing is changed in this enum class.

Comment (jcpitre) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/table/TableStatus.java:1]: I would say it's not essential but it's not bad to have a file dedicated to one enum.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsDescriptor.java
Comment (jcpitre) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsDescriptor.java:1]: Added comments.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsContainer.java
Comment (jcpitre) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsContainer.java:1]: Added class comments.

**Comment by emmambd** [comment]
@jcpitre One note re: the test notice of `unique_location_id_violation`. We won't actually be able to merge this one until the schema for location_groups.txt in #1749 is added, since the rule requires checking location id across all 3 files. 

So there are two options: 

1) Leave the notice is purely for testing purposes and then take it out of this PR. (Moving it to the other issue/PR for #1749) 
2) Add the `location_groups.txt` schema within this PR, which seems to make the PR extra big.

**Comment by emmambd** [comment]
@jcpitre I wouldn't want to release with it in the present state (meaning #1749 would block release, which it already does), but I'm okay with that. False positives are the main focus of this release, not negatives.

**Comment by emmambd** [comment]
@jcpitre I tested duplicate_key with this PR using 
[flex.zip](https://github.com/user-attachments/files/16935573/flex.zip) to try to generate a `duplicate_key` notice. 

It worked! But I realized that we don't have the csvRowNumber lines...because it's not a csv. 
Screenshot 2024-09-09 at 3 30 34 PM

Do we have a plan or need to discuss how to modify these columns for generic notices can affect either a geojson file or a csv file? cc @davidgamez

**Comment by emmambd** [comment]
@emmambd Try to break/generate other generic notices since we weren't expecting `duplicate_key`!

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableDescriptor.java [hunk 1]
```diff
@@ -7,35 +7,18 @@
 import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
 import org.mobilitydata.gtfsvalidator.parsing.CsvHeader;
 
-public abstract class GtfsTableDescriptor<T extends GtfsEntity> {
+public abstract class GtfsTableDescriptor<T extends GtfsEntity> extends GtfsFileDescriptor<T> {
```
Comment (davidgamez) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableDescriptor.java:10]: [suggestion]: rename this class to `GtfsCsvFileDescriptor`.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/notice/GeojsonDuplicateKeyNotice.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Duplicated feature id.
+ *
+ * <p>One of the feature id of the locations.geojson file appears twice or more in the file.
+ */
+@GtfsValidationNotice(
+    severity = ERROR,
+    files = @GtfsValidationNotice.FileRefs(GtfsLocationsSchema.class),
+    sections = @SectionRefs(FILE_REQUIREMENTS))
+public class GeojsonDuplicateKeyNotice extends ValidationNotice {
```
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/notice/GeojsonDuplicateKeyNotice.java:34]: As discussed, we can create this notice more generic renaming it to `JsonDuplicateKeyNotice, adding it the field name, field value and removing references to the featureId.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/JsonFileLoader.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import org.mobilitydata.gtfsvalidator.notice.IOError;
+import org.mobilitydata.gtfsvalidator.notice.NoticeContainer;
+import org.mobilitydata.gtfsvalidator.validator.ValidatorProvider;
+
+public class JsonFileLoader extends TableLoader {
```
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/JsonFileLoader.java:16]: This looks no generic enough to name it as `JsonFileLoader`. I suggest to rename it to GeoJsonFileLoader or make the implementation generic enough to cover any JSON descriptor.

File: processor/notices/src/main/java/org/mobilitydata/gtfsvalidator/processor/notices/NoticesProcessor.java [hunk 4]
```diff
@@ -52,31 +52,35 @@ public SourceVersion getSupportedSourceVersion() {
...
+            GSON.toJson(comments.get(), writer);
+          }
+        } catch (IOException e) {
+          throw new RuntimeException(e);
         }
-      } catch (IOException e) {
-        throw new RuntimeException(e);
       }
+    } catch (Exception e) {
+      System.out.println("Error: " + e);
```
Comment (davidgamez) [inline processor/notices/src/main/java/org/mobilitydata/gtfsvalidator/processor/notices/NoticesProcessor.java:83]: 👁️ Don't forget to revert these changes before merge if they were added to debug. If there are no other changes in this class other than formatting, I suggestion to revert all changes to reduce noise in this already large PR.

File: processor/tests/src/test/java/org/mobilitydata/gtfsvalidator/processor/tests/CurrencyAmountSchemaTest.java [hunk 5]
```diff
@@ -1,84 +0,0 @@
...
-import org.junit.runners.JUnit4;
-import org.mobilitydata.gtfsvalidator.notice.InvalidCurrencyAmountNotice;
-import org.mobilitydata.gtfsvalidator.table.CurrencyAmountTableDescriptor;
-import org.mobilitydata.gtfsvalidator.testing.LoadingHelper;
-import org.mobilitydata.gtfsvalidator.validator.CurrencyAmountCurrencyAmountValidator;
-import org.mobilitydata.gtfsvalidator.validator.ValidatorLoader;
-import org.mobilitydata.gtfsvalidator.validator.ValidatorLoaderException;
-
-@RunWith(JUnit4.class)
-public class CurrencyAmountSchemaTest {
```
Comment (davidgamez) [inline processor/tests/src/test/java/org/mobilitydata/gtfsvalidator/processor/tests/CurrencyAmountSchemaTest.java:34]: 👁️ revert these changes before merge.

File: processor/tests/src/test/java/org/mobilitydata/gtfsvalidator/processor/tests/MixedCaseSchemaTest.java [hunk 6]
```diff
@@ -1,110 +0,0 @@
...
-import org.mobilitydata.gtfsvalidator.notice.MixedCaseRecommendedFieldNotice;
-import org.mobilitydata.gtfsvalidator.table.MixedCaseTest;
-import org.mobilitydata.gtfsvalidator.table.MixedCaseTestTableDescriptor;
-import org.mobilitydata.gtfsvalidator.testing.LoadingHelper;
-import org.mobilitydata.gtfsvalidator.validator.MixedCaseTestMixedCaseValidator;
-import org.mobilitydata.gtfsvalidator.validator.ValidatorLoader;
-import org.mobilitydata.gtfsvalidator.validator.ValidatorLoaderException;
-
-@RunWith(Parameterized.class)
-public class MixedCaseSchemaTest {
```
Comment (davidgamez) [inline processor/tests/src/test/java/org/mobilitydata/gtfsvalidator/processor/tests/MixedCaseSchemaTest.java:36]: 👁️ revert there changes before merge.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java [hunk 7]
```diff
@@ -115,13 +117,11 @@ private <T extends GtfsTableContainer<E>, E extends GtfsEntity> void setCount(
     var table = feedContainer.getTableForFilename(fileName);
     this.counts.put(
         countName,
-        table
-            .map(gtfsTableContainer -> loadUniqueCount(gtfsTableContainer, clazz, idExtractor))
-            .orElse(0));
+        table.map(GtfsContainer -> loadUniqueCount(GtfsContainer, clazz, idExtractor)).orElse(0));
```
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:120]: rename parameter to:
```suggestion
        table.map(gtfsContainer -> loadUniqueCount(gtfsContainer, clazz, idExtractor)).orElse(0));
```

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/AnyTableLoader.java [hunk 8]
```diff
@@ -7,48 +7,42 @@
...
 import org.mobilitydata.gtfsvalidator.parsing.FieldCache;
 import org.mobilitydata.gtfsvalidator.parsing.RowParser;
-import org.mobilitydata.gtfsvalidator.validator.FileValidator;
 import org.mobilitydata.gtfsvalidator.validator.SingleEntityValidator;
 import org.mobilitydata.gtfsvalidator.validator.ValidatorProvider;
 import org.mobilitydata.gtfsvalidator.validator.ValidatorUtil;
 
-public final class AnyTableLoader {
+/** This class loads csv files specifically. */
+public final class AnyTableLoader extends TableLoader {
```
[the remark this answers is not in the document]
Comment (jcpitre) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/table/AnyTableLoader.java:27]: Done

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedContainer.java [hunk 9]
```diff
@@ -18,20 +18,19 @@
 
 import com.google.common.base.Ascii;
 import java.util.*;
-import org.mobilitydata.gtfsvalidator.table.GtfsTableContainer.TableStatus;
 
 /**
  * Container for a whole parsed GTFS feed with all its tables.
  *
- * <p>The tables are kept as {@code GtfsTableContainer} instances.
+ * <p>The tables are kept as {@code GtfsContainer} instances.
```
[the remark this answers is not in the document]
Comment (jcpitre) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedContainer.java:25]: Done

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableDescriptor.java (hunk 1 above)
Comment (jcpitre) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableDescriptor.java:10]: To be logical all TableDescriptor (e.g. CurrentcyAmountTableDescriptor) should also be renamed.

Comment (jcpitre) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableDescriptor.java:10]: This is becoming a massive change. To do things properly I would need to change any reference from Table to File. (e.g. TableLoader becomes FileLoader).
I am thinking of not doing this and having a dedicated PR for that change.

File: processor/notices/src/main/java/org/mobilitydata/gtfsvalidator/processor/notices/NoticesProcessor.java (hunk 4 above)
Comment (jcpitre) [reply processor/notices/src/main/java/org/mobilitydata/gtfsvalidator/processor/notices/NoticesProcessor.java:83]: This try catch was added by you. Was it only for debugging?

Comment (davidgamez) [reply processor/notices/src/main/java/org/mobilitydata/gtfsvalidator/processor/notices/NoticesProcessor.java:83]: Yes, it was due to an exception regarding the null tables, but that is not the case now.

Comment (jcpitre) [reply processor/notices/src/main/java/org/mobilitydata/gtfsvalidator/processor/notices/NoticesProcessor.java:83]: Reverted

**Comment by jcpitre** [comment]
@emmambd This PR is ready for a bit of QA

**Review by emmambd — APPROVED** [review]
Re-tested `duplicate_key` and `missing_required_field` based on the functionality here and in #1749. Working as expected. 

Expected behaviour for 6.0 release (so we can document scope): 

- `duplicate_key` is triggered for `location_groups.txt` and `stops.txt` in cases where there is a duplicate key within the same file. `duplicate_key` is not triggered in `locations.geojson` because we'll be adding a new notice in release 7.0 to cover this use case. `unique_geography_id` is a new notice that will check across all three files, but will not be added until 7.0. 

- `missing_required_field` triggers when `stop_id` is not included in `stop_times.txt` and there are no other geography ids present for the stop time. If `locations.id` or `location_group.id` are included, `missing_required_field` does not trigger. 

@davidgamez @jcpitre I'm not sure if we need the memory usage checks before merging this PR (I don't think so? but will leave you two to discuss!) On the spec side, this looks good to me.

**Review by davidgamez — APPROVED** [review]
LGTM

**Pull request merged**